sdk/attestation/azure-security-attestation/src/attestation_administration_client.cpp

```diff
@@ -0,0 +1,150 @@
+  }
+
+  namespace {
+    std::shared_timed_mutex SharedStateLock;
+  }
+
+  Response<Models::AttestationToken<std::string>>
+  AttestationAdministrationClient::GetAttestationPolicy(
+      AttestationType const& attestationType,
+      GetPolicyOptions const& options,
+      Azure::Core::Context const& context) const
+  {
+    Log::Write(
+        Logger::Level::Informational,
+        std::string("Get Policy for AttestationType:") + attestationType.ToString());
+
+    auto request = AttestationCommonRequest::CreateRequest(
+        m_endpoint,
+        m_apiVersion,
+        HttpMethod::Get,
+        {"policies/" + attestationType.ToString()},
+        nullptr);
+
+    // Send the request to the service.
+    auto response = AttestationCommonRequest::SendRequest(*m_pipeline, request, context);
+
+    // Deserialize the Service response token and return the JSON web token returned by the
+    // service.
+    std::string responseToken = AttestationServiceTokenResponseSerializer::Deserialize(response);
+
+    // Parse the JWT returned by the attestation service.
+    auto resultToken
+        = AttestationTokenInternal<Models::_detail::PolicyResult, PolicyResultSerializer>(
+            responseToken);
+
+    // Validate the token returned by the service. Use the cached attestation signers in the
+    // validation.
+    std::vector<AttestationSigner> const& signers = GetAttestationSigners(context);
+    resultToken.ValidateToken(
+        options.TokenValidationOptions ? options.TokenValidationOptions.Value()
+                                       : this->m_tokenValidationOptions,
+        signers);
+
+    // Extract the underlying policy token from the response.
+    std::string policyTokenValue
+        = static_cast<AttestationToken<Models::_detail::PolicyResult>>(resultToken)
+              .Body.PolicyToken.Value();
+
+    // TPM policies are empty by default, at least in our test instances, so handle the empty policy
+    // token case.
+    auto policyTokenI
+        = AttestationTokenInternal<StoredAttestationPolicy, StoredAttestationPolicySerializer>(
+            policyTokenValue);
+    AttestationToken<StoredAttestationPolicy> policyToken(policyTokenI);
+    std::string returnPolicy;
+    if (policyToken.Body.AttestationPolicy)
+    {
+      std::vector<uint8_t> policyUtf8 = policyToken.Body.AttestationPolicy.Value();
+      returnPolicy = std::string(policyUtf8.begin(), policyUtf8.end());
+    }
+
+    // Construct a token whose body is the policy, but whose token is the response from the
+    // service.
+    auto returnedToken = AttestationTokenInternal<std::string>(responseToken, returnPolicy);
+    return Response<AttestationToken<std::string>>(returnedToken, std::move(response));
+  }
+
+  std::vector<AttestationSigner> const& AttestationAdministrationClient::GetAttestationSigners(
+      Azure::Core::Context const& context) const
+  {
+    std::unique_lock<std::shared_timed_mutex> stateLock(SharedStateLock);
```

Comment:
I am not familiar enough with the implementation detail or requirements from the attestation SDK, so take this question with a grain of salt, and largely out of curiosity.

We have typically avoided taking locks or mutexes in the client libraries. Why is this necessary?

Reply:
It takes several hundred ms to retrieve the attestation signing certificates. If we don't cache them on the client, an attestation client which makes more than one API call to the service is going to burn hundreds of milliseconds of wall time retrieving the exact same data again and again, once for every single call into the service.

This avoids that cost.

Reply:
Thanks! That would be great as a comment in source, to explain the rationale :)